ext/ffi/ffi.c

```diff
@@ -1221,6 +1221,25 @@ static zend_result zend_ffi_cdata_cast_object(zend_object *readobj, zval *writeo
 }
 /* }}} */
 
+static int zend_ffi_should_convert_cdata(void)
+{
+	zend_execute_data *execute_data = EG(current_execute_data);
+
+	if (execute_data
+	 && EX(opline)
+	 && EX(call)
+	 && (EX(opline)->opcode == ZEND_FETCH_OBJ_R
+	  || EX(opline)->opcode == ZEND_FETCH_OBJ_FUNC_ARG
+	  || EX(opline)->opcode == ZEND_FETCH_DIM_R
+	  || EX(opline)->opcode == ZEND_FETCH_DIM_FUNC_ARG)
+	 && EX(call)->func
+	 && EX(call)->func->type == ZEND_INTERNAL_FUNCTION
+	 && EX(call)->func->internal_function.module == &ffi_module_entry) {
```

Comment:
There might be an issue (I didn't test it) when passing a value from an FFI value to one of the FFI functions that are currently _not_ `@prefer-ref`. For example:

```
\FFI::memset($x, $y['z'], 42);
```

If `$y` is a `CData`, `$y['z']` will return another `CData` object to represent the offset. This would throw a type error, at least with `strict_types=1`. Luckily, it doesn't look like there are many such cases. In theory, they could also be solved by accepting `CData` explicitly, and fetching the value internally.

Reply:
Right again. We should check that the result of FETCH_DIM/OBJ used as an operand of SEND or FRAMELESS_CALL and make a final decision taking into account the called function and parameter. The test becomes too complex...

Reply:
I had the same thought. This would be trivial for `FETCH_DIM_R`. It only uses flags in `extended_value`. `FETCH_OBJ_R` contains the cache slot offset in `extended_value`. However, it also stores `ZEND_FETCH_REF` (`1`), so we already remove it through a bitwise op when fetching the cache slot. We could either use `2`, as the cache slot offset is a multiple of `4` on 32-bit, and `8` on 64-bit. Hence, there shouldn't be any performance overhead.

Reserving an engine flag for an optional extension does feel a bit unfortunate. That said, we likely could still use higher flags (`1 << 31`) when we need to.